vdr/resolvers.go

```diff
@@ -0,0 +1,126 @@
+package vdr
+
+import (
+	"crypto"
+	"fmt"
+	"github.com/nuts-foundation/go-did"
+	vdr "github.com/nuts-foundation/nuts-node/vdr/types"
+	"hash/crc32"
+	"time"
+)
+
+type EndpointResolver interface {
+	Resolve(subject did.DID, endpointType string) ([]did.Service, error)
+}
+
+func NewVDREndpointResolver(registry vdr.VDR) EndpointResolver {
+	return &vdrEndpointResolver{registry: registry}
+}
+
+type vdrEndpointResolver struct {
+	registry vdr.VDR
+}
+
+func (v vdrEndpointResolver) Resolve(subject did.DID, endpointType string) ([]did.Service, error) {
+	// TODO: should we provide validAt?
+	document, _, err := v.registry.Resolve(subject, nil)
+	if err != nil {
+		return nil, err
+	}
+	var result []did.Service
+	for _, service := range document.Service {
+		if service.Type == endpointType {
+			result = append(result, service)
+		}
+	}
+	return result, nil
+}
+
+type KeyPurpose string
+
+const (
+	AuthenticationKeyPurpose  = "authentication"
+	AssertionMethodKeyPurpose = "assertionMethod"
+)
+
+const errKeyNotFoundFmt = "key not found (holder=%s,purpose=%s,validAt=%s)"
+
+type DIDKeyResolver interface {
+	ResolvePublicKey(keyID string, purpose KeyPurpose, validAt *time.Time) (crypto.PublicKey, error)
+	ResolveKeyID(holder did.DID, purpose KeyPurpose, validAt *time.Time) (string, error)
+}
+
+func NewVDRDIDKeyResolver(registry vdr.VDR) DIDKeyResolver {
+	return &vdrKeyResolver{registry: registry}
+}
+
+type vdrKeyResolver struct {
+	registry vdr.VDR
+}
+
+func (r vdrKeyResolver) ResolvePublicKey(keyID string, purpose KeyPurpose, validAt *time.Time) (crypto.PublicKey, error) {
+	holder, err := did.ParseDID(keyID)
+	if err != nil {
+		return nil, err
+	}
+	keys, err := findKeys(r.registry, *holder, purpose, validAt)
+	if err != nil {
+		return nil, err
+	}
+	for _, key := range keys {
+		if key.ID.String() == keyID {
+			var publicKey crypto.PublicKey
+			err = key.JWK().Raw(&publicKey)
+			return publicKey, err
+		}
+	}
+
+	return nil, nil
+}
+
+func (r vdrKeyResolver) ResolveKeyID(holder did.DID, purpose KeyPurpose, validAt *time.Time) (string, error) {
+	keys, err := findKeys(r.registry, holder, purpose, validAt)
+	if err != nil || len(keys) == 0 {
+		return "", err
+	}
+	return keys[0].ID.String(), nil
+}
+
+func findKeys(registry vdr.VDR, holder did.DID, purpose KeyPurpose, validAt *time.Time) ([]did.VerificationRelationship, error) {
```

Comment:
Shouldn't this be available on did.Document?

Reply:
Removed altogether.